lib/http_aws_sigv4.c

```diff
@@ -66,6 +66,102 @@ static void sha256_to_hex(char *dst, unsigned char *sha, size_t dst_l)
   }
 }
 
+/*
+ * Decide in an encoding-independent manner whether a character in an
+ * URL must be escaped. The same criterion must be used in strlen_uri()
+ * and strcpy_uri().
+ */
+static bool urlchar_needs_escaping(int c, bool in_query)
+{
+  return !((c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') ||
+           (c >= '0' && c <= '9') || c == '-' || c == '.' || c == '_'
+           || c == '~' ||
+           (!in_query && c == '/') ||
+           (in_query && c == '&'));
+}
+
+/*
+ * strlen_uri() returns the length of the given URI
+ * this fucntion is a modified version of strlen_url in urlapi.c
+ */
+static size_t encode_query_len(const char *url, bool in_query)
+{
+  const unsigned char *ptr;
+  size_t newlen = 0;
+  int count_equal = 0;
+
+  for(ptr = (unsigned char *)url; *ptr; ptr++) {
+    switch (*ptr) {
+    case '=':
+      if(!in_query || count_equal++)
+        newlen += 2;
+      newlen++;
+      break;
+    case '%':
+      newlen++;
+      break;
+    case '&':
+      count_equal = 0;
+      /* FALLTHROUGH */
+    default:
+      if(urlchar_needs_escaping(*ptr, in_query))
+        newlen += 2;
+      newlen++;
+    }
+  }
+  return newlen;
+}
+
+/* encode_query_cpy() copies a url to a output buffer and URL-encodes
+ * the URI acordingly
+ * this fucntion is a modified version of strcpy_url in urlapi.c
+ */
+static void encode_query_cpy(char *optr, const char *iptr, bool in_query)
+{
+  /* we must add this with whitespace-replacing */
+  int count_equal = 0;
+
+  for(;
+      *iptr;         /* until zero byte */
+      iptr++) {
+
+    switch(*iptr) {
+    case '=':
+      if(!in_query || count_equal++) {
+        msnprintf(optr, 4, "%%%02X", *iptr);
+        optr += 3;
+      }
+      else
+        *optr++=*iptr;
+      break;
+    case '%':
+      *optr++ = *iptr;
+      if(iptr[1]) {
+        iptr++;
+        *optr++ = (char)TOUPPER(*iptr);
+        if(*iptr != '%' && iptr[1]) {
```

Comment:
Why do you compare `*iptr` with `'%'`?

Reply:
so if you have an encoded URL, containing '%2a' for example I can use '%2A' in the canonical header.  I detect '%' so I can decode/encode URL

Reply:
What about for example `%41` present in the original URL. Shouldn't that rather get decoded to `A` in a canonical URL?

Reply:
@outscale-mgo, but you've detected '%' while `case '%'` yet, haven't you?

Reply:
> What about for example `%41` present in the original URL. Shouldn't that rather get decoded to `A` in a canonical URL?

ok  I'll do that

Reply:
****

> @outscale-mgo, but you've detected '%' while `case '%'` yet, haven't you?

yes, because %% is deferent from %01 which represent an ASCII character, anyway I'm going to change the algo to make a full decode -> encode.